contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -192,96 +225,100 @@ type EphemeralContainer struct {
+The entire v1.Container used in the request will be serialized and stored as a
+runtime annotation. The value of `TargetContainerName` will be stored as a
+runtime label. Persisting this data in the runtime means it will survive kubelet
+restarts.
+
+At least for the Docker runtime, this is [an intended use of docker
+labels](https://docs.docker.com/engine/userguide/labels-custom-metadata/#value-guidelines).
+Docker does not document the maximum length of labels in its API. Empirically,
+it supports up to the 64K constraint of the docker client's `bufio.Scanner`
+size. Because the container spec may be examined in security sensitive contexts
+like admission control, we will conservatively limit the size of the spec to 32K
+and add a 32K minimum label length test to runtime qualification.
+
+`EphemeralContainer` is populated by the kubelet in the same way as regular
+container statuses. This is sent to the API server and displayed by `kubectl
+describe pod`.
 
 ### Creating Debug Containers
 
-1.  `kubectl` invokes the exec API as described in the preceding section.
+1.  `kubectl` invokes the new API as described in the preceding section.
 1.  The API server checks for name collisions with existing containers, performs
-    admission control and proxies the connection to the kubelet's
-    `/exec/$NS/$POD_NAME/$CONTAINER_NAME` endpoint.
-1.  The kubelet instructs the Runtime Manager to create a Debug Container.
-1.  The runtime manager uses the existing `startContainer()` method to create a
-    container in an existing pod. `startContainer()` has one modification for
-    Debug Containers: it creates a new runtime label (e.g. a docker label) that
-    identifies this container as a Debug Container.
-1.  After creating the container, the kubelet schedules an asynchronous update
-    of `PodStatus`. The update publishes the debug container status to the API
-    server at which point the Debug Container becomes visible via `kubectl
-    describe pod`.
-1.  The kubelet will upgrade the connection to streaming and attach to the
-    container's console.
-
-Rather than performing the implicit attach the kubelet could return success to
-the client and require the client to perform an explicit attach, but the
-implicit attach maintains consistent semantics across `/exec` rather than
-varying behavior based on parameters.
+    validation, admission control and proxies the connection to the kubelet's
+    `/ephemeralContainers/$NS/$POD_NAME` endpoint.
+1.  The kubelet request handler opens an error channel and signals the pod's
+    sync worker with `UpdatePodOptions` that include the `EphemeralContainer` in
+    a new field and a callback in the existing
+    `UpdatePodOptions.OnCompleteFunc`.
+    1.  The pod sync worker runs the existing `syncPod()` with a new
+        `SyncPodType` of `SyncPodDebug`.
+    1.  The request handler blocks (with timeout) on receiving an error from
```

Comment:
what steps occur within this timeout? pulling the image? how is the timeout selected here?

Reply:
Yes, the container is started, including pulling the image if necessary. I was just going to make the timeout flag configurable. I'd prefer to improve the reliability of the `OnCompleteFunc` for `SyncPodDebug` to making the timeout more configurable. I've described this a little bit in the alpha notes. It should be easy to get to the point where the timeout is rarely used.

Reply:
Oh, also I updated the doc to reflect this.

Reply:
I'm trying to determine whether an API request to add an ephemeral container should be considered a long running request on the API side. Pulling the image within the lifetime of an API request is a new direction here.

Also, if pulling happens prior to the kubelet persisting the ephemeral container spec in the pod status, does that mean the kubelet gets a single shot at pulling and never reattempts or reports status on failure?

Reply:
Oh, interesting. Yes, by design we want a failure in starting the ephemeral container to be a permanent failure. There will never be any reattempts, but it should report the failure back to the user, and in the pod's event log.

For example, image pull failures return a ErrImagePull to the user. I'm not sure of the current kubelet behavior for a really slow image pull, but this would be the same for the Ephemeral Container since it uses the same code.

So if len(image_pull_secs) > len(api_request_timeout_secs) then the user will get error("start ephemeral container timed out") even though the ephemeral container will eventually be created (or not). If the Ephemeral Container is eventually created, it will be available for manual attach.

Reply:
> by design we want a failure in starting the ephemeral container to be a permanent failure

that seems fine

> There will never be any reattempts, but it should report the failure back to the user

I'm having difficulty understanding how this works when combined with the image pull happening within the timeout. Would a pull failure still result in a pod status update? Would it make more sense for the subresource POST to just update pod status with the requested ephemeral container, the kubelet notice that via normal pod watch events, and update pod status with the results of attempting to start the ephemeral container?

That would allow the client side implementation to look like:
1. submit ephemeral subresource request, get back immediate success or error
2. on success, monitor pod status until ephemeral container is running
3. attach